### DISCUSSION TITLE
Too slow to load data from a large dataset due to the retrieving indicies

### DISCUSSION THREAD

**Discussion by Nokimann** [pyg-team/pytorch_geometric#6454] [Q&A]
The number of data is over 1,000,000 (~40GB).

I checked the custom `get` function in `Dataset` class is not a problem.
The problem is `indices` function in `Dataset` class that causes this lazy loading.

```
    def indices(self) -> Sequence:
        return range(self.len()) if self._indices is None else self._indices
```

Loading from dataloader of the large dataset
```
%%time
for i in range(16):
    data = dataset[i]

CPU times: user 32.8 s, sys: 3.69 s, total: 36.5 s
Wall time: 36.5 s
```

Loading directly from files
```
%%time
for i in range(16):
    data = torch.load(filenames[i])

CPU times: user 10 ms, sys: 451 µs, total: 10.5 ms
Wall time: 9.6 ms
```

Just doing `len(dataset)`
```
%%time
for i in range(16):
    len(dataset)

CPU times: user 33.6 s, sys: 3.67 s, total: 37.3 s
Wall time: 37.3 s
```

I just replaced `self.indices()[idx]` to just `idx` in `__getitem__` function, and solved temporarily.
I'm not sure this causes some other problems, so hope to find the best solution...
```
    def __getitem__(
        self,
        idx: Union[int, np.integer, IndexType],
    ) -> Union['Dataset', Data]:
        r"""In case :obj:`idx` is of type integer, will return the data object
        at index :obj:`idx` (and transforms it in case :obj:`transform` is
        present).
        In case :obj:`idx` is a slicing object, *e.g.*, :obj:`[2:5]`, a list, a
        tuple, or a :obj:`torch.Tensor` or :obj:`np.ndarray` of type long or
        bool, will return a subset of the dataset at the specified indices."""
        if (isinstance(idx, (int, np.integer))
                or (isinstance(idx, Tensor) and idx.dim() == 0)
                or (isinstance(idx, np.ndarray) and np.isscalar(idx))):

            data = self.get(idx) # data = self.get(self.indices()[idx])
            data = data if self.transform is None else self.transform(data)
            return data

        else:
            return self.index_select(idx)
```

**Comment by rusty1s** [comment]
How do you currently define your `def len` function? I think it is very important that you define that in a fast way, e.g., not doing a `glob` to get the number of files. Other than that, we could consider caching the length. Would that resolve your concerns?

**Reply by Nokimann** [reply]
I didn't concern about that because that is not my crucial problem, so `len(self)` fucntion is the same definition as in the [example](https://pytorch-geometric.readthedocs.io/en/latest/notes/create_dataset.html#creating-larger-datasets)

```
    def len(self):
        return len(self.processed_file_names)
```

But, we can define a internal variable inside a user-defined large dataset to call in `def len(self)`.

**Reply by rusty1s** [reply]
Yes, but how is `self.processed_file_names` defined?

**Reply by Nokimann** [reply]
```
    @property
    def processed_file_names(self) -> str:
        return [f'data_{idx}.pt' for idx in self.file_indices]
```

Oh! This is problem. So, the large dataset must be defined with the internal variable of filenames
